baseapp/baseapp.go

```diff
@@ -936,25 +975,52 @@ func (app *BaseApp) runTx(mode execMode, txBytes []byte) (gInfo sdk.GasInfo, res
 // and DeliverTx. An error is returned if any single message fails or if a
 // Handler does not exist for a given message route. Otherwise, a reference to a
 // Result is returned. The caller must not commit state if an error is returned.
-func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, msgsV2 []protov2.Message, mode execMode) (*sdk.Result, error) {
+func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, msgsV2 []protov2.Message, mode execMode, isNonAtomic bool) (*sdk.Result, error) {
 	events := sdk.EmptyEvents()
 	var msgResponses []*codectypes.Any
 
+	txSuccess := false
+	msgCtx := ctx // This context may be replaced on each message for non-atomic execution.
+
 	// NOTE: GasWanted is determined by the AnteHandler and GasUsed by the GasMeter.
 	for i, msg := range msgs {
 		if mode != execModeFinalize && mode != execModeSimulate {
 			break
 		}
 
+		// When dealing with non-atomic execution, we want each message to have its own context
+		var writeFn func()
+		if isNonAtomic {
+			msgCtx, writeFn = ctx.CacheContext()
+		}
+
 		handler := app.msgServiceRouter.Handler(msg)
 		if handler == nil {
 			return nil, errorsmod.Wrapf(sdkerrors.ErrUnknownRequest, "can't route message %+v", msg)
 		}
 
 		// ADR 031 request type routing
-		msgResult, err := handler(ctx, msg)
+		msgResult, err := handler(msgCtx, msg)
 		if err != nil {
-			return nil, errorsmod.Wrapf(err, "failed to execute message; message index: %d", i)
+			space, code, logInfo := errorsmod.ABCIInfo(err, mode != execModeFinalize)
```

Comment:
mode must equal execModeFinalize in message execution. I would just apply true here.